PR for feedback

ddxxdd-code opened the pull request.

ddxxdd-code committed:
Add instrumentations for openj9. first version, without test.

ddxxdd-code committed:
Fix wrong type for region log

ddxxdd-code committed:
Compile OK version

ddxxdd-code committed:
Version verified to work on simple tests

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3778,6 +3779,26 @@ void TR::CompilationInfo::stopCompilationThreads()
+   // check if there is any compilation logged in the global list, if so, dump the contents
+   if (J9::Options::_collectRegionLog && J9::Options::_compilationRegionLogFileName)
+      {
+      std::FILE *regionLogOutFile = fopen(J9::Options::_compilationRegionLogFileName, "w+");
```
Why do you need `std::`?

ddxxdd-code replied:
fopen by default will return TR::FILE but functions that takes the file descriptor to print to file in RegionLog are designed not in TR namespace. Should I put RegionLog into TR namespace? Or use TR::FILE in the function definitions there?

AlexeyKhrabrov replied:
I think `RegionLog` should go into the `TR::` namespace (along with `Region`) regardless of this.

ddxxdd-code replied:
I tried to use TR:: namespce, but it turned out that fopen, fprintf, fclose only accpets std::FILE. When I use FILE* (TR::FILE *)here, the compiler complains. And I found other examples of using fopen, return value is handled in FILE (not TR::FILE) in other files in OpenJ9.
Compiler complaints:
<img width="717" alt="Screen Shot 2023-01-13 at 00 24 05" src="https://user-images.githubusercontent.com/65443542/212449916-aaffb828-645b-4a90-aa92-5beca262ee10.png">
Other cases:
<img width="795" alt="Screen Shot 2023-01-13 at 23 08 31" src="https://user-images.githubusercontent.com/65443542/212449985-5f2623b1-139a-4ef8-9f07-a3942a9bd453.png">


Another thing I discovered is that I'll need J9::IO::fopen to get TR::FILE, is this the correct way to do? And if I do it this way, I feel I have to find a counterpart in OMR that does the same and takes TR::FILE as well to substitute all of my fprintfs.

ddxxdd-code replied:
Currently I keep this only one std::FILE and compilation is ok. And I haven't seen a use case of J9::IO::fopen() besides its definition or any file that includes the header file for J9IO.hpp.

ddxxdd-code replied:
Oh, probably all of them includes IO.hpp, I'll try using that instead.

ddxxdd-code replied:
Changed to using J9::IO::func in OpenJ9 and TR::IO::func in OMR. Compilation OK

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3778,6 +3779,26 @@ void TR::CompilationInfo::stopCompilationThreads()
       }
 #endif /* defined(J9VM_OPT_JITSERVER) */
+   // check if there is any compilation logged in the global list, if so, dump the contents
+   if (J9::Options::_collectRegionLog && J9::Options::_compilationRegionLogFileName)
```
`TR::Options` (in other places in this file too).

AlexeyKhrabrov commented on runtime/compiler/env/SystemSegmentProvider.cpp:
```diff
@@ -57,10 +69,22 @@ J9::SystemSegmentProvider::SystemSegmentProvider(size_t defaultSegmentSize, size
    {
+
+   // put this segment provider if needed
+   if (_recordRegions && _regionBytesAllocated > J9::Options::_minMemoryUsageCollectRegionLog)
```
`TR::Options`.

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3778,6 +3779,26 @@ void TR::CompilationInfo::stopCompilationThreads()
+      std::FILE *regionLogOutFile = fopen(J9::Options::_compilationRegionLogFileName, "w+");
+      if (!regionLogOutFile)
+         {
+         fprintf(stderr, "error in opening file %s for dumping region logs\n", J9::Options::_compilationRegionLogFileName);
```
Either `perror(fileName);` or try to match other places in this file.

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -8312,6 +8339,7 @@ TR::CompilationInfoPerThreadBase::compile(J9VMThread * vmThread,
    vmThread->omrVMThread->vmState = oldState;
    vmThread->jitMethodToBeCompiled = NULL;
 
+
```
This should be reverted.

AlexeyKhrabrov commented on runtime/compiler/control/J9Options.cpp:
```diff
@@ -19,7 +19,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
-
+ 
```
This should be reverted.

AlexeyKhrabrov commented on runtime/compiler/control/J9Options.cpp:
```diff
@@ -735,6 +741,10 @@ TR::OptionTable OMR::Options::_feOptions[] = {
         TR::Options::setJitConfigNumericValue, offsetof(J9JITConfig, codeCachePadKB), 0, "F%d (KB)"},
    {"codetotal=",              "C<nnn>\ttotal code memory limit, in KB",
         TR::Options::setJitConfigNumericValue, offsetof(J9JITConfig, codeCacheTotalKB), 0, "F%d (KB)"},
+   // set this to enable collection of region log
```
Why do some of the options you added not have `NOT_IN_SUBSET`?

AlexeyKhrabrov commented on runtime/compiler/env/SystemSegmentProvider.cpp:
```diff
@@ -22,13 +22,25 @@
+#include <tuple>
+
+uint32_t J9::SystemSegmentProvider::_globalCompilationSequenceNumber = 0;
+std::vector<std::tuple<uint32_t, size_t, RegionLog *>> *J9::SystemSegmentProvider::_globalCompilationsList = new std::vector<std::tuple<uint32_t, size_t, RegionLog *>>;
```
`std::tuple` is not the best choice for this, just define a struct.

AlexeyKhrabrov commented on runtime/compiler/env/SystemSegmentProvider.cpp:
```diff
@@ -22,13 +22,25 @@
+#include <tuple>
+
+uint32_t J9::SystemSegmentProvider::_globalCompilationSequenceNumber = 0;
+std::vector<std::tuple<uint32_t, size_t, RegionLog *>> *J9::SystemSegmentProvider::_globalCompilationsList = new std::vector<std::tuple<uint32_t, size_t, RegionLog *>>;
```
Why do these need to use `malloc` rather than the persistent allocator?

ddxxdd-code replied:
Oh, yes. How to access the persistent allocator here?

AlexeyKhrabrov replied:
Same as anywhere else, e.g. `TR::Compiler->persistentAllocator()`.

AlexeyKhrabrov commented on runtime/compiler/env/SystemSegmentProvider.cpp:
```diff
@@ -57,10 +69,22 @@ J9::SystemSegmentProvider::SystemSegmentProvider(size_t defaultSegmentSize, size
       }
    _systemBytesAllocated += _systemSegmentSize;
+
+   // log this segment provider
```
Can't this run in multiple threads concurrently?

ddxxdd-code replied:
Then I'll need a monitor to guard this step. Any example for using a monitor here in openj9?

AlexeyKhrabrov replied:
You definitely don't need a lock, I meant that you might need atomic increment. There is also a possibility that the segment provider is created while the compilation queue lock is held, I'll have to check that.

ddxxdd-code replied:
I remember a segment provider is created for only one compilation thread, so it will not be shared by others? Besides, this line is original from the segment provider code in OpenJ9.

AlexeyKhrabrov replied:
My comment is about getting the compilation sequence number (lines 74-75).

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3778,6 +3779,26 @@ void TR::CompilationInfo::stopCompilationThreads()
+         {
+         for (auto &compilation : *J9::SystemSegmentProvider::_globalCompilationsList)
+            {
+            fprintf(regionLogOutFile, "Compilation %d, Memory usage: %zu\n", std::get<0>(compilation), std::get<1>(compilation));
```
Isn't the compilation sequence number unsigned?

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3778,6 +3779,26 @@ void TR::CompilationInfo::stopCompilationThreads()
+         {
+         for (auto &compilation : *J9::SystemSegmentProvider::_globalCompilationsList)
+            {
+            fprintf(regionLogOutFile, "Compilation %d, Memory usage: %zu\n", std::get<0>(compilation), std::get<1>(compilation));
```
Do you print the name of the method being compiled and the optimization level anywhere?

ddxxdd-code replied:
Emmm, that'sanother of my concern, I plan to make the thing a struct instead of the tuple and include these information, but I don't know how to access these fields.

AlexeyKhrabrov replied:
Through the compilation queue entry ("method being compiled") when the segment provider is created. I thought I showed you that at some point? I think you can get the "IL generator method details" or something like that from the method being compiled, and the `J9Method` from there, then `J9ROMMethod` and `J9ROMClass` from there, and eventually the class name and the method name and signature. Let me know if you have trouble figuring this out and I'll give you the actual details (when I have time to look more into the code).

ddxxdd-code replied:
Oh that complicated and lengthy way of getting it. I fear it's overwritten by newly added code.

AlexeyKhrabrov replied:
All the APIs involved in doing this are supposed to be pretty stable, so I don't think that's a concern. I think it's important to include the name and the optimization level of the compiled method in the output.

ddxxdd-code replied:
Now it's printed in the first line of a compilation. The first line consists of compilation number (size_t), total memory usage (size_t), method being compiled (char */%s), and compilation optimization level (size_t).

ddxxdd-code committed:
Update code according to comments

Use struct substitute std::tuple in compilation lists
Use TR::FILE instead of std::FILE for IO
Use of TR::Options instead of J9::Options for newly introduced options
Use perror/TR_ASSERT to substitute fprint error message
Revert extra newlines introduced
Set all newly introduced options to be NOT_IN_SUBSET as they should work for all
Use persistent allocator for global compilations list
Fix possible multi-threading problem on compilation Number by using atomic addition
Print method compiled and optimization level when dump

Signed-off-by: Dedong Xie <dedong.xie@mail.utoronto.ca>

ddxxdd-code committed:
Remove commented code

Remove commented code

Signed-off-by: Dedong Xie <dedong.xie@mail.utoronto.ca>

ddxxdd-code committed:
Add monitor to addition to global log list

Signed-off-by: Dedong Xie <dedong.xie@mail.utoronto.ca>
